RIOT-OS/RIOT
# drivers/nrf24l01p: rewrite: adaptation for gnrc and netdev

### Contribution description

This PR contributes a rewrite of the nrf24l01+ driver. The old driver
was outdated and not compatible with RIOT´s `gnrc` and `netdev` interface.
The nrf24l01+ transceiver is a very cheap 2.4 GHz transceiver which has the special
feature of 6 receiver addresses and 6 Rx data pipes.

Features:

- Do not attempt to acquire the SPI bus for every register SPI- read/write operation
- `gnrc` and `netdev` adaptation
- Possibility to configure, whether to use `ShockBurst` or `Enhanced ShockBurst` as L2 protocol
- L2 broadcast, via the dedicated address `ff:ff:ff:ff:ff`
- ipv6 over 6LoWPAN
- An application `nrf24ctl` to configure a transceiver
  (Configuration via `ifconfig` is also possible, but only considers data pipe 0)
- A pseudomodule `nrf24l01p_diagnostics` that prints a snapshot of all registers and device information

Unfortunately, the transceiver only has a very limited FIFO size of 32 bytes, so even with
6LoWPAN header compression, there is not much left for the application layer.

A strange property of the nrf24l01+ is that the recipient of a frame does not know
the source address. Instead, the sender temporary changes it´s Rx address to the
same address, it has just sent a frame to, and the recipient sends the ACK to his own address.
To support 6LoWPAN, the source address is mandatory. The sender includes it´s actual source address, if `gnrc_sixlowpan` is included in `USEMODULES`.
However, the full 32 bytes FIFO size can still be used, if 6LoWPAN is not required.

Sender and recipient must be configured to have the same address width.
Once set, the address width cannot be changed.
For `ShockBurst`, only static payload lengths are supported and sender and recipient must
be configured to expect the same payload length.

### Testing procedure

I tested with 2 nucleo-f767zi. Take a look at
`tests/driver_nrf24l01p/README.md` to see which params need to be set.

I tested:

`txtsnd 6 e7:e7:e7:e7:e7 RIOT-OS`

`ping6 -c 8 -s 32 fe80::e7e7:00ff:fee7:e7e7`

Output of `ping6`:
```
2019-11-10 11:38:57,600 # --- fe80::e7e7:00ff:fee7:e7e7 PING statistics ---
2019-11-10 11:38:57,605 # 8 packets transmitted, 8 packets received, 0% packet loss
2019-11-10 11:38:57,609 # round-trip min/avg/max = 188.012/188.106/188.223 ms
```

gcoap:

You must remove "`USEMODULE += gnrc_netdev_default`" from `examples/gcoap/makefile`
and for that include "`USEMODULE += gnrc_netif`" and "`USEMODULE += nrf24l01p`".

`coap get -c fe80::e7e7:00ff:fee7:e7e7%7 5683 /riot/board`


### Issues/PRs references
None
